File: src/core/strip.ts
import type { Comment, StripOptions } from '../types.js';

interface Range {
  start: number;
  end: number;
}

function mergeRanges(ranges: readonly Range[]): Range[] {
  const sorted = [...ranges].sort((a, b) => a.start - b.start || a.end - b.end);
  const merged: Range[] = [];
  for (const range of sorted) {
    const last = merged[merged.length - 1];
    if (last && range.start <= last.end) {
      if (range.end > last.end) last.end = range.end;
      continue;
    }
    merged.push({ ...range });
  }
  return merged;
}

function isHorizontalWhitespace(char: string | undefined): boolean {
  return char === ' ' || char === '\t';
}

const BYTE_ORDER_MARK = '\uFEFF';

function expand(source: string, range: Range): Range {
  let start = range.start;
  while (start > 0 && isHorizontalWhitespace(source[start - 1])) start -= 1;
  const consumedLeading = start < range.start;

  const atLineStart =
    start === 0 || source[start - 1] === '\n' || (start === 1 && source[0] === BYTE_ORDER_MARK);

  let cursor = range.end;
  while (isHorizontalWhitespace(source[cursor])) cursor += 1;
  const atLineEnd = cursor >= source.length || source[cursor] === '\n' || source[cursor] === '\r';

  if (!atLineStart) {
    return consumedLeading ? { start, end: range.end } : { start, end: cursor };
  }

  if (!atLineEnd) return { start: range.start, end: cursor };

  let end = cursor;
  if (source[end] === '\r') end += 1;
  if (source[end] === '\n') end += 1;
  return { start, end };
}

export function stripComments(
  source: string,
  comments: readonly Comment[],
  options: StripOptions = {}
): string {
  if (comments.length === 0) {
    return options.collapseBlankLines ? collapseBlankLines(source) : source;
  }

  const ranges = mergeRanges(comments.map(comment => expand(source, comment)));

  let out = '';
  let cursor = 0;
  for (const range of ranges) {
    if (range.start < cursor) continue;
    out += source.slice(cursor, range.start);
    cursor = range.end;
  }
  out += source.slice(cursor);

  return options.collapseBlankLines ? collapseBlankLines(out) : out;
}

export function collapseBlankLines(source: string): string {
  return source.replace(/[ \t]+(\r?\n)/g, '$1').replace(/(\r?\n){3,}/g, '$1$1');
}

File: packages/python/src/commentless/strip.py
from __future__ import annotations

import re

from .types import Comment

BYTE_ORDER_MARK = "﻿"

_TRAILING_WHITESPACE = re.compile(r"[ \t]+(\r?\n)")
_BLANK_RUN = re.compile(r"(\r?\n){4,}")


def _char_at(source: str, index: int) -> str | None:
    if index < 0 or index >= len(source):
        return None
    return source[index]


def _is_horizontal_whitespace(char: str | None) -> bool:
    return char in (" ", "\t")


def _merge_ranges(ranges: list[tuple[int, int]]) -> list[tuple[int, int]]:
    merged: list[tuple[int, int]] = []
    for start, end in sorted(ranges):
        if merged and start <= merged[-1][1]:
            if end > merged[-1][1]:
                merged[-1] = (merged[-1][0], end)
            continue
        merged.append((start, end))
    return merged


def _expand(source: str, start: int, end: int) -> tuple[int, int]:
    cursor_start = start
    while cursor_start > 0 and _is_horizontal_whitespace(_char_at(source, cursor_start - 1)):
        cursor_start -= 1
    consumed_leading = cursor_start < start

    at_line_start = (
        cursor_start == 0
        or _char_at(source, cursor_start - 1) == "\n"
        or (cursor_start == 1 and source[:1] == BYTE_ORDER_MARK)
    )

    cursor = end
    while _is_horizontal_whitespace(_char_at(source, cursor)):
        cursor += 1
    at_line_end = cursor >= len(source) or source[cursor] in ("\n", "\r")

    if not at_line_start:
        return (cursor_start, end) if consumed_leading else (cursor_start, cursor)

    if not at_line_end:
        return start, cursor

    stop = cursor
    if _char_at(source, stop) == "\r":
        stop += 1
    if _char_at(source, stop) == "\n":
        stop += 1
    return cursor_start, stop


def strip_comments(
    source: str,
    comments: list[Comment],
    *,
    collapse_blank_lines_option: bool = False,
) -> str:
    if not comments:
        return collapse_blank_lines(source) if collapse_blank_lines_option else source

    ranges = _merge_ranges([_expand(source, c.start, c.end) for c in comments])

    parts: list[str] = []
    cursor = 0
    for start, end in ranges:
        if start < cursor:
            continue
        parts.append(source[cursor:start])
        cursor = end
    parts.append(source[cursor:])

    out = "".join(parts)
    return collapse_blank_lines(out) if collapse_blank_lines_option else out


def collapse_blank_lines(source: str) -> str:
    return _BLANK_RUN.sub(r"\1\1\1", _TRAILING_WHITESPACE.sub(r"\1", source))
